File: simple_auth_flutter/ios/Classes/WebAuthenticator.m
//
//  WebAuthenticator.m
//  Runner
//
//  Created by James Clancey on 6/6/18.
//  Copyright © 2018 The Chromium Authors. All rights reserved.
//

#import "WebAuthenticator.h"
#import <Flutter/Flutter.h>

@implementation WebAuthenticator
-(id)initFromDictionary:(NSDictionary *)data_
{
    self = [super init];
    NSLog(@"Authenticator created");
    if(self) {
        self.identifier = [data_ valueForKey:@"identifier"];
        self.allowsCancel = [[data_ valueForKey:@"allowsCancel"] boolValue];
        self.useEmbeddedBrowser =[[data_ valueForKey:@"useEmbeddedBrowser"] boolValue];
        self.initialUrl = [NSURL URLWithString:[data_ valueForKey:@"initialUrl"]];
        self.redirectUrl = [NSURL URLWithString:[data_ valueForKey:@"redirectUrl"]];
        self.title = [data_ valueForKey:@"title"];
    }
    return self;
}
-(void)cancel{
    _eventSink(@{
                 @"identifier" : self.identifier,
                 @"url" : @"canceled"
                 });
}
-(void)foundToken{
    self.isCompleted = YES;
    _onTokenFound();
}
-(void) checkUrl:(NSURL *)url forceComplete:(BOOL)force
{
    _eventSink(@{
                 @"identifier" : self.identifier,
                 @"url" : url.absoluteString,
                 @"forceComplete" : force ? @"true" : @"false"
                 });
}
-(void) failed:(NSString *)error{
    _eventSink(@{
                 @"identifier" : self.identifier,
                 @"url" : @"error",
                 @"description": error.description,
                 @"forceComplete" : @"true"
                 });
}
@end

File: simple_auth_flutter/android/src/main/java/clancey/simpleauth/simpleauthflutter/WebAuthenticator.java
package clancey.simpleauth.simpleauthflutter;

import java.net.URI;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.common.MethodCall;


public class WebAuthenticator {

    public WebAuthenticator(MethodCall call) throws Exception
    {
        identifier = call.argument("identifier");
        initialUrl = call.argument("initialUrl").toString();
        redirectUrl = new URI(call.argument("redirectUrl").toString());
        title = call.argument("title");
        allowsCancel = Boolean.parseBoolean((String)call.argument("allowsCancel"));
        isCompleted = Boolean.parseBoolean((String)call.argument("isCompleted"));
        useEmbeddedBrowser = Boolean.parseBoolean((String)call.argument("useEmbeddedBrowser"));


    }
    public String identifier;
    public String initialUrl;
    public URI redirectUrl;
    public String title;
    public boolean allowsCancel;
    public boolean isCompleted;
    public boolean useEmbeddedBrowser;
    public EventChannel.EventSink eventSink;
    public void checkUrl(final String url, final boolean forceComplete)
    {
        eventSink.success( new HashMap<String, String>() {{
            put("identifier",identifier);
            put("url",url);
            put("forceComplete", forceComplete ? "true":"false");

        }});
    }
    List<CompleteNotifier> listeners = new ArrayList<CompleteNotifier>();
    public void foundToken()
    {
        isCompleted = true;
        for (CompleteNotifier l : listeners)
            l.onComplete();
    }

    public void addListener(CompleteNotifier listener)
    {
        listeners.add(listener);
    }

    public void clearListeners()
    {
        listeners.clear();
    }
    public void cancel()
    {
        eventSink.success( new HashMap<String, String>() {{
            put("identifier",identifier);
            put("url","canceled");
        }});
    }

    public void failed(final String error)
    {
        eventSink.success( new HashMap<String, String>() {{
            put("identifier",identifier);
            put("url","error");
            put("description", error);
            put("forceComplete", "true");
        }});
    }
    public class CompleteNotifier
    {
        public void onComplete()
        {

        }
    }
}
